protocol/decoupled-hardfork-features.md

```diff
@@ -0,0 +1,308 @@
+# Purpose
+
+The purpose of this design document is to propose a mechanism for decoupling
+individual features from their hardforks in the OP Stack. This will enhance
+flexibility in feature activation, enabling granular control for testing,
+custom chain configurations, and phased rollouts while maintaining
+coordination across the superchain. The design aims to improve modularity
+without introducing excessive complexity, ultimately supporting more robust
+development and deployment practices of OP Stack software.
+
+# Background
+
+The OP Stack relies on the `RollupConfig` to specify chain parameters,
+including hardfork activation timestamps (e.g., `canyonTime`, `deltaTime`).
+These configurations are centralized in the [`superchain-registry`][scr]
+repository, as TOML files under `superchain/configs/`. For chains that are
+part of the superchain, activations are inherited from a shared timestamp
+called the `superchain_time` to ensure synchronized upgrades.
+
+Currently, when a hardfork timestamp is reached in OP Stack client software,
+all features within that hardfork (e.g., EIPs, protocol changes) activate
+simultaneously. Bundling features in this way simplifies coordination but
+limits flexibility, such as enabling individual features for testing or
+delaying specific ones without forking the config.
+
+In the past, contributors of the OP Stack have had issues surrounding
+hardfork features. For example, long before the Isthmus hardfork, the
+Operator Fee Feature was [proposed][proposed] and approved. The feature
+was considered to be a part of the upcoming Isthmus hardfork. Rust
+implementations of OP Stack components went ahead and implemented the
+feature. For months, the feature sat idle. After it was eventually merged,
+there were issues in the execution client changes for the feature. Removing
+implementations of the feature was considered, but the effort to remove it
+was not seen as worthwhile. After a while, the feature was finally
+complete. It effectively alone delayed the hardfork by a month or so.
+
+This could have been avoided by being able to implement the feature using
+a feature-specific toggle in the `RollupConfig`. The feature toggle would
+be toggled on in the [`superchain-registry`][scr] once the feature is ready.
```

Comment:
Putting the feature toggle in the rollup config is one way, but we could have equally feature toggled it by using `if IsOperatorFeeEnabled()` in place of `if IsIsthmus()` and while we expect operator fee to ship with Isthmus it's implementation could just delegate to `IsIsthmus`.  Though in the case of operator fees there were also smart contract changes involved which would also have needed to be feature toggled (and notably those don't have access to the rollup config).

Reply:
Do I understand correctly that the main change you're suggesting here is not to feature toggle in the SCR, but instead just feature toggle in the `RollupConfig` implementations in clients?

I actually think this is a great idea. My one concern though is if we think a few years down the road, without centralizing on a set of features, we could see clients meaningfully diverge as they implement features in different ways. Maybe this is a feature and not a bug (no pun intended), but it's something to keep in mind. Being explicit + centralized about feature names and what hardfork they target could prevent a lot of tech debt build up and potential chain split surface area.